text/0000-ptr-meta.md

```diff
@@ -0,0 +1,320 @@
+impl VTable {
+    /// Returns the size of the type associated with this vtable.
+    pub fn size(&self) -> usize { ... }
+
+    /// Returns the alignment of the type associated with this vtable.
+    pub fn align(&self) -> usize { ... }
+
+    /// Returns the size and alignment together as a `Layout`
+    pub fn layout(&self) -> alloc::Layout {
+        unsafe {
+            alloc::Layout::from_size_align_unchecked(self.size(), self.align())
+        }
+    }
+
+    /// Drops the value pointed at by `data` assuming it has the type
+    /// associated with this vtable.
+    ///
+    /// Behaviour is Undefined if the type isn’t correct or the pointer
+    /// is invalid to reinterpret as `&mut TheType`.
+    pub unsafe fn drop_in_place(&self, data: *mut ()) { ... }
+}
+```
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The status quo is that code (such as linked in [Motivation]) that requires this functionality
+needs to transmute to and from `std::raw::TraitObject`
+or a copy of it (to be compatible with Stable Rust).
+Additionally, in cases where constructing the data pointer
+requires knowing the alignment of the concrete type,
+a dangling pointer such as `0x8000_0000_usize as *mut ()` needs to be created.
+It is not clear whether `std::mem::align_of(&*ptr)` with `ptr: *const dyn SomeTrait`
+is Undefined Behavior with a dangling data pointer.
+
+A [previous iteration][2579] of this RFC proposed a `DynTrait`
+that would only be implemented for trait objects like `dyn SomeTrait`.
+There would be no `Metadata` associated type, `&'static VTable` was hard-coded in the trait.
+In addition to being more general
+and (hopefully) more compatible with future custom DSTs proposals,
+this RFC resolves the question of what happens
+if trait objects with super-fat pointers with multiple vtable pointers are ever added.
+(Answer: they can use a different metadata type like `[&'static VTable; N]`.)
+
+`VTable` could be made generic with a type parameter for the trait object type that it describes.
+This would avoid forcing that the size, alignment, and destruction pointers
+be in the same location (offset) for every vtables.
+But keeping them in the same location is probaly desirable anyway to keep code size
+
+[2579]: https://github.com/rust-lang/rfcs/pull/2579
+
+
+# Prior art
+[prior-art]: #prior-art
+
+A previous [Custom Dynamically-Sized Types][cdst] RFC was postponed.
+[Internals thread #6663][6663] took the same ideas
+and was even more ambitious in being very general.
+Except for `VTable`’s methods, this RFC proposes a subset of what that thread did.
+
+[cdst]: https://github.com/rust-lang/rfcs/pull/1524
+[6663]: https://internals.rust-lang.org/t/pre-erfc-lets-fix-dsts/6663
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* The name of `Pointee`. [Internals thread #6663][6663] used `Referent`.
+
+* The location of `VTable`. Is another module more appropriate than `std::ptr`?
```

Comment:
and should it be called `Dictionary` instead? ("type class dictionary")

Reply:
Big -1 to calling it Dictionary since this typically means a key-value map (example: [C#](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=netframework-4.7.2), [Swift](https://docs.swift.org/swift-book/LanguageGuide/CollectionTypes.html), [Python](https://docs.python.org/3/library/stdtypes.html#typesmapping)). 

Furthermore, here in Rust the `VTable` is implemented as an array of function pointers, not a HashMap (unlike e.g. Python where it is really implemented as a `dict`), so calling it Dictionary obscures the alleged complexity.